[Bug] Possible Race Condition in joern-slice leading to overflowdb errors

When importing and slicing code with `joern-slice`, sometimes a `overflowdb.SchemaViolationException` is thrown.
This was observed for multiple TypeScript projects.
As mentioned in #2439, this seems to be occurring non-deterministically.
During testing, an error occurred in ca. 1 out of 20 runs.

**To Reproduce**
Steps to reproduce the behavior:
1. Extract the attached example project and run
```
joern-slice -m Usages ./example/ -o ./
```
[example.zip](https://github.com/joernio/joern/files/11158323/example.zip)

2. If CPG creation succeeds, `rm example.json` and repeat step 1. 
3. Observe the following error:
```java
Exception in thread "main" overflowdb.SchemaViolationException: OUT edge with label AST to an adjacent METHOD_RETURN is mandatory, but not defined for this METHOD node with id=590
	at io.shiftleft.codepropertygraph.generated.nodes.MethodDb.methodReturn(Method.scala:645)

```

<details><summary>Full Stack Trace</summary>
<p>

```java
[+] Applying default overlays
WARNING: sun.reflect.Reflection.getCallerClass is not supported. This will impact performance.
Temporary CPG has been successfully generated at /tmp/joern-slice2972118312677447146.bin
Exception in thread "main" overflowdb.SchemaViolationException: OUT edge with label AST to an adjacent METHOD_RETURN is mandatory, but not defined for this METHOD node with id=590
	at io.shiftleft.codepropertygraph.generated.nodes.MethodDb.methodReturn(Method.scala:645)
	at io.shiftleft.codepropertygraph.generated.nodes.Method.methodReturn(Method.scala:204)
	at io.shiftleft.codepropertygraph.generated.traversal.MethodTraversalExtGen$.$anonfun$methodReturn$1(Method.scala:35)
	at scala.collection.Iterator$$anon$9.next(Iterator.scala:577)
	at scala.collection.Iterator$$anon$9.next(Iterator.scala:577)
	at scala.collection.IterableOps.headOption(Iterable.scala:231)
	at scala.collection.IterableOps.headOption$(Iterable.scala:229)
	at overflowdb.traversal.Traversal.headOption(Traversal.scala:16)
	at io.joern.joerncli.slicing.UsageSlicing$.$anonfun$trackUsage$6(UsageSlicing.scala:166)
	at scala.collection.Iterator$$anon$10.nextCur(Iterator.scala:587)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:601)
	at scala.collection.IterableOps.headOption(Iterable.scala:231)
	at scala.collection.IterableOps.headOption$(Iterable.scala:229)
	at overflowdb.traversal.Traversal.headOption(Traversal.scala:16)
	at io.joern.joerncli.slicing.UsageSlicing$.exprToObservedCall$1(UsageSlicing.scala:165)
	at io.joern.joerncli.slicing.UsageSlicing$.$anonfun$trackUsage$14(UsageSlicing.scala:181)
	at scala.collection.Iterator$$anon$10.nextCur(Iterator.scala:587)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:601)
	at scala.collection.immutable.List.prependedAll(List.scala:155)
	at scala.collection.IterableOnceOps.toList(IterableOnce.scala:1251)
	at scala.collection.IterableOnceOps.toList$(IterableOnce.scala:1251)
	at overflowdb.traversal.Traversal.toList(Traversal.scala:16)
	at io.joern.joerncli.slicing.UsageSlicing$.partitionInvolvementInCalls$1(UsageSlicing.scala:181)
	at io.joern.joerncli.slicing.UsageSlicing$.trackUsage(UsageSlicing.scala:209)
	at io.joern.joerncli.slicing.UsageSlicing$.$anonfun$calculateUsageSlice$3(UsageSlicing.scala:45)
	at scala.collection.parallel.AugmentedIterableIterator.flatmap2combiner(RemainsIterator.scala:123)
	at scala.collection.parallel.AugmentedIterableIterator.flatmap2combiner$(RemainsIterator.scala:121)
	at scala.collection.parallel.immutable.ParVector$ParVectorIterator.flatmap2combiner(ParVector.scala:70)
	at scala.collection.parallel.ParIterableLike$FlatMap.leaf(ParIterableLike.scala:1038)
	at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:52)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:97)
	at scala.collection.parallel.Task.tryLeaf(Tasks.scala:55)
	at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:49)
	at scala.collection.parallel.ParIterableLike$FlatMap.tryLeaf(ParIterableLike.scala:1034)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.compute(Tasks.scala:152)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.compute$(Tasks.scala:148)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$AWSFJTWrappedTask.compute(Tasks.scala:304)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.util.NoSuchElementException: next on empty iterator
	at scala.collection.Iterator$$anon$19.next(Iterator.scala:966)
	at scala.collection.Iterator$$anon$19.next(Iterator.scala:964)
	at scala.collection.Iterator$$anon$7.next(Iterator.scala:528)
	at overflowdb.traversal.Traversal.next(Traversal.scala:22)
	at io.shiftleft.codepropertygraph.generated.nodes.MethodDb.methodReturn(Method.scala:640)
	... 43 more

```

</p>
</details> 

More rarely, the following error is observed:
```java
Exception in thread "main" java.lang.NullPointerException
	at overflowdb.NodeDb.startIndex(NodeDb.java:664)
	at overflowdb.NodeDb.createAdjacentNodeIteratorByOffSet(NodeDb.java:595)
	at io.shiftleft.codepropertygraph.generated.nodes.MethodDb.astOut(Method.scala:638)
	at io.shiftleft.codepropertygraph.generated.nodes.MethodDb.methodReturn(Method.scala:640)
	at io.shiftleft.codepropertygraph.generated.nodes.Method.methodReturn(Method.scala:204)
	at io.shiftleft.codepropertygraph.generated.traversal.MethodTraversalExtGen$.$anonfun$methodReturn$1(Method.scala:35)
	at scala.collection.Iterator$$anon$9.next(Iterator.scala:577)
	at scala.collection.Iterator$$anon$9.next(Iterator.scala:577)
	at scala.collection.IterableOps.headOption(Iterable.scala:231)
	at scala.collection.IterableOps.headOption$(Iterable.scala:229)
	at overflowdb.traversal.Traversal.headOption(Traversal.scala:16)
	at io.joern.joerncli.slicing.UsageSlicing$.$anonfun$trackUsage$6(UsageSlicing.scala:166)
	at scala.collection.Iterator$$anon$10.nextCur(Iterator.scala:587)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:601)
	at scala.collection.IterableOps.headOption(Iterable.scala:231)
	at scala.collection.IterableOps.headOption$(Iterable.scala:229)
	at overflowdb.traversal.Traversal.headOption(Traversal.scala:16)
	at io.joern.joerncli.slicing.UsageSlicing$.exprToObservedCall$1(UsageSlicing.scala:165)
	at io.joern.joerncli.slicing.UsageSlicing$.$anonfun$trackUsage$14(UsageSlicing.scala:181)

[...]
```

**Expected behavior**
CPG creation should always succeed and must not fail non-deterministically.

**Desktop (please complete the following information):**
 - OS: Linux 5.15.0-1030-gcp 37-Ubuntu
 - Joern Version: 1.1.1563
 - javac 11.0.18

--- a/joern-cli/src/main/scala/io/joern/joerncli/slicing/UsageSlicing.scala
+++ b/joern-cli/src/main/scala/io/joern/joerncli/slicing/UsageSlicing.scala
@@ -9,6 +9,7 @@ import overflowdb.traversal.Traversal
 
 import java.util.regex.Pattern
 import scala.collection.parallel.CollectionConverters.ImmutableIterableIsParallelizable
+import scala.util.Try
 
 object UsageSlicing {
 
@@ -163,7 +164,7 @@ object UsageSlicing {
       // Not sure how we can get the return type unless it's typescript or we can resolve the callee?
       val returnType = baseCall.argumentOut
         .flatMap {
-          case x: Call => x.callee(resolver).methodReturn.typeFullName.headOption
+          case x: Call => Option(Try(x.callee(resolver).methodReturn.typeFullName.head).getOrElse(null))
           case _       => None
         }
         .headOption

--- a/joern-cli/src/main/scala/io/joern/joerncli/slicing/UsageSlicing.scala
+++ b/joern-cli/src/main/scala/io/joern/joerncli/slicing/UsageSlicing.scala
@@ -165,10 +165,7 @@ object UsageSlicing {
       val returnType = baseCall.argumentOut
         .flatMap {
           case x: Call =>
-            Try(x.callee(resolver).methodReturn.typeFullName.head) match {
-              case Failure(_) => None
-              case Success(t) => Option(t)
-            }
+            Try(x.callee(resolver).methodReturn.typeFullName.head).toOption
           case _ => None
         }
         .headOption

--- a/joern-cli/src/main/scala/io/joern/joerncli/slicing/UsageSlicing.scala
+++ b/joern-cli/src/main/scala/io/joern/joerncli/slicing/UsageSlicing.scala
@@ -9,7 +9,7 @@ import overflowdb.traversal.Traversal
 
 import java.util.regex.Pattern
 import scala.collection.parallel.CollectionConverters.ImmutableIterableIsParallelizable
-import scala.util.Try
+import scala.util.{Failure, Success, Try}
 
 object UsageSlicing {
 
@@ -164,8 +164,12 @@ object UsageSlicing {
       // Not sure how we can get the return type unless it's typescript or we can resolve the callee?
       val returnType = baseCall.argumentOut
         .flatMap {
-          case x: Call => Option(Try(x.callee(resolver).methodReturn.typeFullName.head).getOrElse(null))
-          case _       => None
+          case x: Call =>
+            Try(x.callee(resolver).methodReturn.typeFullName.head) match {
+              case Failure(_) => None
+              case Success(t) => Option(t)
+            }
+          case _ => None
         }
         .headOption
         .getOrElse("ANY")

